### PR TITLE
Ignore empty documents.

### PR DESCRIPTION
**Pull request by ah8ad3**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Ignore empty documents.
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kubectl-validate/issues/109
#### Special notes for your reviewer:
I am not sure ignoring would effect other things or not (we don't have any failing test), please point out if it would.
#### Does this PR introduce a user-facing change?

```release-note
Ignore empty documents and accept them as a valid file.
```

### CODE REVIEW

**Comment by alexzielenski** [comment]
Looking thru the code for this PR i see we already have logic to skip empty YAML documents, I am curious how that bug persists...


https://github.com/kubernetes-sigs/kubectl-validate/blob/8b66d4cdabca2b502a29bc4b48c6e8ea4d820dd4/pkg/cmd/validate.go#L261-L263

**Comment by alexzielenski** [comment]
@ah8ad3 could you add a test case for the linked issue and investigate why the existing logic for this does not work correctly for that case before we consider removing error checking logic

File: pkg/utils/yaml.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func SplitYamlDocuments(fileBytes Document) ([]Document, error) {
 func IsEmptyYamlDocument(document Document) bool {
 	for _, line := range strings.Split(string(document), "\n") {
 		line := strings.TrimSpace(line)
-		if line != "" && !strings.HasPrefix(line, "#") {
+		if line != "" && !strings.HasPrefix(line, "#") && !strings.HasPrefix(line, "---") {
```
Comment (alexzielenski) [inline pkg/utils/yaml.go:33]: Rather than `HasPrefix` I think we want to check that `line == "---"`

File: pkg/utils/yaml.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ func SplitYamlDocuments(fileBytes Document) ([]Document, error) {
 func IsEmptyYamlDocument(document Document) bool {
```
Comment (alexzielenski) [inline pkg/utils/yaml.go:30]: Could you add some godoc explaining that this function returns true for comment-only single documents, and strings with multiple documents where all docs are comment-only.

**Review by alexzielenski — COMMENTED** [review]
Thanks for the test cases! Just one comment about HasPrefix.

It surprised me that a document with just `---` would include `---` in the output.

Usually the terminator is stripped by `SplitYAMLDocuments`, but reading over the implementation, it looks like the terminator is included in cases where the doc is empty: e.g. `---\n` yields a single document output with contents `---` where I would expect two empty documents. Which seems like a bug to me, but unlikely to be fixed upstream.

I think this workaround is good, `IsEmptyYamlDocument` returns true if every line is either a document terminator or a comment.

File: pkg/utils/yaml.go (hunk 2 above)
Comment (ah8ad3) [reply pkg/utils/yaml.go:30]: > comment-only single documents

But it'll work for files like this too:
1: 
```
# test doc
# test doc
---
# test doc
# test doc
---
# test doc
# test doc
```
2: 
```
---
---
```

I think gendoc should be like this:
This function validate emptiness of yaml if it only contains comment-only documents or strings with multiple documents.

If we want to make the second example invalid we should change the logic here.

Comment (alexzielenski) [reply pkg/utils/yaml.go:30]: Yes those two cases are covered by the second clause of my suggestion:

> Could you add some godoc explaining that this function returns true for comment-only single documents, ***and strings with multiple documents where all docs are comment-only***.

I'm not particular about the specific wording, as long as it mentions it works on multiple document yaml files, and comments are still considered empty.

Comment (ah8ad3) [reply pkg/utils/yaml.go:33]: Updated, PTAL

**Comment by ah8ad3** [comment]
Commits squashed.

**Review by alexzielenski — APPROVED** [review]
/lgtm
/approve

**Comment by alexzielenski** [comment]
CI is unhappy with a comment it seems

**Comment by ah8ad3** [comment]
It was a odd one, my IDE didn't warn me that :|. Gofmt was not happy with `,` at the end of the line.
It should be fixed now. @alexzielenski .

**Pull request merged**